Support scripting animation in python through PyO3

AzurIce opened the pull request.

AzurIce committed:
feat: basic python API

AzurIce committed:
clean up

AzurIce committed:
refactor: rewrite color system to use `color` instead of `bevy_color`

AzurIce committed:
chore(dep): update deps

AzurIce committed:
style: cargo fmt and cargo clippy

AzurIce committed:
moved pyo3 implementation to ranimpy package

AzurIce committed:
chore: gh deploy only when push to main

AzurIce committed:
style: cargo fmt and cargo clippy

AzurIce committed:
basic attempt to interpret script in rust successed

AzurIce committed:
fixed module name

AzurIce committed:
style: cargo clippy and cargo fmt

AzurIce committed:
cleanup

AzurIce committed:
cargo fmt

AzurIce committed:
feat: read timelines from module, then build and render them all to output/<name> dir

AzurIce committed:
use clap for cli

AzurIce committed:
refactor: module structure

AzurIce committed:
feat: created pyo3 bindings for Transformable methods

shenjackyuanjie commented:
关于 pyo3 在实际使用的时候可能会遇到的环境问题 (比如运行在一个启动了venv的终端里，但是并不使用）
可以参考 [这里的代码](https://github.com/shenjackyuanjie/icalingua-bridge-bot/blob/fe411ba1371205a6606af1bf935317c8c0b67989/ica-rs/src/py/mod.rs#L487-L531)

shenjackyuanjie commented on packages/ranim-cli/src/main.rs:
```diff
@@ -0,0 +1,277 @@
+
+    #[test]
+    fn test_downcast() -> PyResult<()> {
+        pyo3::append_to_inittab!(ranimpy_module);
+        pyo3::prepare_freethreaded_python();
+        Python::with_gil(|py| {
+            let sys = PyModule::import(py, "sys")?;
+
+            let executable = sys.getattr("executable")?;
+            let path = sys.getattr("path")?;
+            let version = sys.getattr("version")?;
+            // if args.len() == 2 {
+            //     let venv_dir = Path::new(&args[1]);
+            let site_packages_path = dunce::canonicalize("../../.venv/Lib/site-packages").unwrap();
+            path.call_method1("append", (site_packages_path.to_str().unwrap(),))
+                .unwrap();
+            // }
+            println!("pyo3 sys.executable: {}", executable);
+            println!("pyo3 sys.path: {}", path);
+            println!("pyo3 sys.version: {}", version);
```
这里的环境处理应该将 173 的 `pyo3::prepare_freethreaded_python();`
替换为手动使用 ffi 直接对 `PyConfig` 进行设置
现在这样的实现并不能实现所需的功能

shenjackyuanjie commented on packages/ranim-cli/src/main.rs:
```diff
@@ -0,0 +1,277 @@
+    }
+
+    #[test]
+    fn test_python_module_methods() -> PyResult<()> {
+        pyo3::append_to_inittab!(ranimpy_module);
+        pyo3::prepare_freethreaded_python();
+        Python::with_gil(|py| {
+            let sys = PyModule::import(py, "sys")?;
+
+            let executable = sys.getattr("executable")?;
+            let path = sys.getattr("path")?;
+            let version = sys.getattr("version")?;
+            // if args.len() == 2 {
+            //     let venv_dir = Path::new(&args[1]);
+            let site_packages_path = dunce::canonicalize("../../.venv/Lib/site-packages").unwrap();
+            path.call_method1("append", (site_packages_path.to_str().unwrap(),))
+                .unwrap();
+            // }
+            println!("pyo3 sys.executable: {}", executable);
+            println!("pyo3 sys.path: {}", path);
+            println!("pyo3 sys.version: {}", version);
```
这里同上

AzurIce committed:
docs(design): record ECS architecture

Design: *D0002, -D0001

AzurIce committed:
refactor: replace store and render graph with ecs

AzurIce committed:
chore: update flakes

AzurIce committed:
chore: update deps

AzurIce committed:
chore: update egui deps

AzurIce committed:
feat: experimental embedded-python scripting support (ranimpy)

The ranim binary now hosts a Python interpreter (pyo3) and can execute
animation scripts directly, instead of ranim being packaged as a Python
library:

    cargo run -p ranim-cli --features python -- render script.py [scenes...]

- packages/ranimpy: bindings for items (VItem/Square/Circle/SvgItem/
  CameraFrame), animations (type-erased Anim over AnimSequence, sequence/
  stack/hold/forward/at, rate functions), palettes, and the
  @ranimpy.scene decorator with name/output_dir/clear_color metadata.
  Scenes registered in a script are rendered through the existing
  SceneConstructor/render pipeline.
- ranim-cli: new optional `python` feature; a positional .py argument to
  `render` switches to script mode (preview reports unsupported for now).
- examples/py: basic.py and composable_choreography.py (ported from the
  Rust example; rendered output verified frame-by-frame).
- flake.nix: add python3 (pyo3 build-time interpreter) and ffmpeg to the
  dev shell; python3 for crane dependency builds.

shenjackyuanjie commented:
近况如何（

AzurIce commented:
> 近况如何（

最近在超级推进其他部分，或许scripting这块后面和editor放在一起做了w

目前基于dylib的热重载方案体验还可以
